### PR TITLE
Added CsvSpecs and readCsv for CsvSpecs.

### PR DESCRIPTION
**Pull request by devinrsmith**
Related to https://github.com/deephaven/deephaven-core/issues/654

### CODE REVIEW

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+            throw new ParserException("Long millis is greater than max instant");
+        }
+        return instant;
+    }
+
+    private static Instant parseAsEpochMicros(Parser<Long> longParser, Instant minInstant, Instant maxInstant,
+            String value) {
+        final long epochMicro = longParser.parse(value);
+        final long epochSecond = Math.floorDiv(epochMicro, 1_000_000);
+        final int nanoAdj = (int) Math.floorMod(epochMicro, 1_000_000);
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java:349]: This is missing a `* 1_000`

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+     * @param max the maximum instant to infer
+     * @return the epoch milli and micro parsers
+     *
+     * @see #epochMilliAndMicroParsers21stCentury(Parser)
+     */
+    public static List<Parser<Instant>> epochMilliAndMicroParsers(Parser<Long> longParser, Instant min, Instant max) {
+        final long minMillis = min.toEpochMilli();
+        final long maxMillis = max.toEpochMilli();
+        final long minMicros = minMillis * 1_000L;
+        if (maxMillis >= minMicros) {
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java:139]: Also validate that min is less than (or equal) max.

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+        if (minInstant != null && instant.isBefore(minInstant)) {
+            throw new ParserException("Long millis is less than min instant");
+        }
+        if (maxInstant != null && instant.isAfter(maxInstant)) {
+            throw new ParserException("Long millis is greater than max instant");
+        }
+        return instant;
+    }
+
+    private static Instant parseAsEpochMicros(Parser<Long> longParser, Instant minInstant, Instant maxInstant,
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java:345]: Should we add nanos?

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                Parser.FLOAT)
+                .build();
+    }
+
+    /**
+     * The parsers, in preference-based order.
+     *
+     * @return the parsers
+     */
+    public abstract List<Parser<?>> parsers();
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java:210]: In some cases, we know that if one parser can parse a particular input, several other parsers also can (e.g. short implies int or long). We ought to try to be aware of this in order to short-circuit parsing in those cases.

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+     * Uses the default {@link #onNullParser()}.
+     *
+     * <p>
+     * Note: the byte and float parsers are only relevant when the appropriate {@link #limitToType(Type)} is invoked.
+     *
+     * @return the standard inference
+     */
+    public static InferenceSpecs standard() {
+        return builder().addParsers(
+                Parser.INSTANT,
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java:145]: Should INSTANT_DB be here?

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+                parsers.get(0),
+                parsers.get(1),
+                Parser.SHORT,
+                Parser.INT,
+                Parser.LONG,
+                Parser.DOUBLE,
+                Parser.BOOL,
+                Parser.CHAR,
+                Parser.STRING,
+                Parser.BYTE,
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java:200]: I'm not in love with making BYTE ignored like this, but you may have a good point about it

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+            final String item = values.next();
+            if (item != null) {
+                allNull = false;
+                candidates.removeIf(parser -> !parser.isParsable(item));
+            }
+        }
+        if (allNull) {
+            return Optional.ofNullable(onNullParser());
+        }
+        return candidates.stream().findFirst();
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java:272]: That is handled by having String parser in inference.

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+            final Iterator<ParserException> it = exceptions.iterator();
+            final ParserException first = it.next();
+            while (it.hasNext()) {
+                first.addSuppressed(it.next());
+            }
+            throw first;
+        }
+    }
+
+    private static boolean parseBool(String value) {
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java:311]: If we need to create more parsers in the future, I think we can.

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java (hunk 3 above)
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/Parser.java:345]: Done.

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java (hunk 5 above)
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java:145]: I'd like to get away from old instant formats... that said, our *default* inference is `standardTimes()` which includes it...

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java (hunk 6 above)
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/InferenceSpecs.java:200]: We might actually want to not parse to SHORT as well... I'm not sure...

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/CsvSpecs.java [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+        @Override
+        public boolean hasNext() {
+            return it.hasNext();
+        }
+
+        @Override
+        public String next() {
+            CSVRecord next = it.next();
+            String stringValue = next.get(index);
+            // treating empty string as null
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/CsvSpecs.java:390]: I said "yes" earlier - now, I think "no"... if a user wants empty strings, they can do an updateView on the column. Thoughts?

File: DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java [hunk 10]
```diff
@@ -558,6 +562,58 @@ public static Table readHeaderlessCsv(File file, String format, StatusCallback p
...
+    /**
+     * Creates an in-memory table from {@code url} according to the {@code specs}.
+     *
+     * @param url the url
+     * @param specs the csv specs
+     * @return the table
+     * @throws IOException if an I/O exception occurs
+     */
+    @ScriptApi
+    public static Table readCsv(URL url, CsvSpecs specs) throws IOException {
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java:613]: I've added an InputStream version now. I'd like to make it easier for users to opt-in or be explicit about their compression format. You were pretty critical of "not double-buffering" - and I think the logic right now for opening a potentially compressed file does its own buffering...

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/CsvSpecs.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (rcaudy) [reply DB/src/main/java/io/deephaven/db/tables/utils/csv/CsvSpecs.java:367]: I have a love/hate relationship with the idea that we rely on update(View) after the fact to correct read issues. On the one hand, it's super flexible, on the other hand it's an extra step/complexity.

File: DB/src/main/java/io/deephaven/db/tables/utils/CsvHelpers.java [hunk 11]
```diff
@@ -575,6 +598,110 @@ public static QueryTable readCsv2(InputStream is) throws IOException {
...
+     * Creates an in-memory table from {@code stream} according to the {@code specs}. The {@code stream} will be closed
+     * upon return.
+     *
+     * @param stream the stream
+     * @param specs the csv specs
+     * @return the table
+     * @throws IOException if an I/O exception occurs
+     */
+    @ScriptApi
+    public static Table readCsv(InputStream stream, CsvSpecs specs) throws IOException {
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/CsvHelpers.java:701]: We should either deprecate, delete, or re-implement the old ones in terms of the new ones.

File: DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java [hunk 12]
```diff
@@ -33,6 +33,7 @@
 
 import java.net.URL;
 import java.nio.file.Path;
+import java.nio.file.Paths;
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java:36]: Unused.

File: DB/src/main/java/io/deephaven/db/tables/utils/CsvHelpers.java [hunk 13]
```diff
@@ -463,49 +418,16 @@ private static StatusCallback checkStatusCallback(StatusCallback progress, boole
...
-     *        passed explicitly as null, a StatusCallback instance will be created to log progress to the current
-     *        logger.
-     * @return a Deephaven Table object
-     * @throws IOException if the InputStream cannot be read
-     */
-    @ScriptApi
-    public static Table readCsv(InputStream is, final String format, StatusCallback progress) throws IOException {
-        final StatusCallback lProgress = checkStatusCallback(progress, true);
-        return readCsvInternal(is, format, false, lProgress);
+        final CsvSpecs specs = CsvSpecs.fromLegacyFormat(format);
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/CsvHelpers.java:426]: Please group these methods thematically by input type, or by spec/no-spec with an input type ordr.

File: DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java [hunk 14]
```diff
@@ -501,10 +490,18 @@ public static Table readHeaderlessCsv(String filePath) throws IOException {
...
      * @return a Deephaven Table object
      * @throws IOException if the file cannot be read
+     * @deprecated See {@link CsvHelpers#readCsv(String, CsvSpecs)}
      */
     @ScriptApi
+    @Deprecated
     public static Table readHeaderlessCsv(String filePath, Collection<String> header) throws IOException {
-        return readHeaderlessCsv(new File(filePath), null, null, header);
+        final List<Selectable> views = new ArrayList<>();
+        int index = 0;
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java:499]: Oh no! A bug! Use 1-based indexing and actually test it this time! ;-)

File: DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java [hunk 15]
```diff
@@ -575,7 +531,7 @@ public static Table readHeaderlessCsv(File file, String format, StatusCallback p
     @ScriptApi
     public static void writeCsv(Table source, boolean compressed, String destPath, String... columns)
             throws IOException {
-        writeCsv(source, compressed, destPath, false, columns);
+        writeCsv(source, compressed, destPath, NULLS_AS_EMPTY_DEFAULT, columns);
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/TableTools.java:534]: See JavaDoc; can we get rid of "DB " everywhere?

File: DB/src/main/java/io/deephaven/db/tables/utils/csv/CsvSpecs.java [hunk 16]
```diff
@@ -118,6 +119,19 @@ public static CsvSpecs headerless(TableHeader header) {
         return builder().hasHeaderRow(false).header(header).build();
     }
 
+    public static CsvSpecs fromLegacyFormat(String format) {
+        if (format == null || "TRIM".equals(format)) {
+            return CsvSpecs.csv();
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/tables/utils/csv/CsvSpecs.java:124]: This needs to set `trim(true)`.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**